testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509BrowserLoginSubjectAltNameEmailTest.java

```diff
@@ -38,14 +42,26 @@
 public class X509BrowserLoginSubjectAltNameEmailTest extends AbstractX509AuthenticationTest {
 
     @Page
+    @PhantomJSBrowser
     protected AppPage appPage;
 
     @Page
+    @PhantomJSBrowser
     protected X509IdentityConfirmationPage loginConfirmationPage;
 
     @Page
+    @PhantomJSBrowser
     protected LoginPage loginPage;
 
+    @Drone
+    @PhantomJSBrowser
+    private WebDriver phantomJS;
+
+    @Before
+    public void replaceTheDefaultDriver() {
+        replaceDefaultWebDriver(phantomJS);
```

Comment:
@hmlnarik @vmuzikar This part is a bit tricky and maybe you guys know better way...

One of the features in Graphene is to use a custom Web Driver for specific pages (see [manual](http://arquillian.org/arquillian-graphene/#browser-instantiation)). This is why I had to add `@PhantomJSBrowser` annotation on each page object. However, all our pages use injected Web Driver ([see the code here](https://github.com/keycloak/keycloak/blob/8b628bda266ac5a459acd848085c450523911ef9/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/AbstractPage.java#L40)). As the injected Web Driver (the one we inject into `AbstractPage`) doesn't use any annotations, an `HTMLUnit` Driver gets injected. From there, it's very easy to confuse Graphene, because we use two different drivers to navigate things on the same page object. This often (but not always) leads to an exception.

I noticed that we used to use `DroneUtils.addWebDriver` to solve similar problems but this doesn't solve the problem entirely, since only a small portion of our pages use `DroneUtils.getCurrentDriver`. In the vast majority of the cases, we use injected Web Driver.

So in order to make sure I switch the driver to Phantom JS completely, I had to:

* Add `@PhantomJSBrowser` annotation to page objects for Graphene.
* Add proper driver to `DroneUtils`.
* Iterate over all page objects in test class and replace the driver previously injected by Arquillian.

If this is the first (and only) time you see this kind of problem, I guess it's OK to leave it like that. But if we struggle with this more often, I think we need do something about it. My guess is either to remove injected Web Driver into page objects (and force using `DroneUtils.getCurrentDriver` instead) or develop some extension to force Arquillian to inject proper driver where we want.

Reply:
I overally agree with the PR and with a way to have the X509 tests executed by default. So from my PoV, I am approving.

Just a note that in the meantime, I sent PR for some additional X509 feature https://github.com/keycloak/keycloak/pull/5937 and I think that in my PR, I may have a conflict with you - I renamed the class "X509BrowserLoginSubjectAltNameEmailTest" to "X509BrowserLoginSubjectAltNameTest" as I needed more tests for SubjectAlternativeName (not just specific to email) and adding additional class looked to me as an unecessary overhead... I don't mind if your PR is merged first and I will then rebase on top of that :)

I am leaving the low-level details regarding WebDriver to clarify with @vmuzikar , @hmlnarik or someone else. Just a note that we have annotation @Different used in some tests (for example SSOTest) and it looks that it's trying to solve the similar problem to yours - even if it is a bit different as you always need the PhantomJSDriver. But it looks that the approach with @Different has similar problem you mentioned - the "driver" injected inside pages is not correct.
I am just pointing that in case if it makes sense to do something and "merge" both approaches? But I am not sure if it would simplify things or if it will be rather more complicated. Just asking and leaving to you guys :)

Reply:
@vmuzikar mentioned that he'll consult this issue with Drone/Graphene expert. Maybe there's a better solution.